### PR TITLE
Support 'multiple' option for item dropdowns

### PR DESCRIPTION
**Pull request by AdrienClairembault**
The `$itemtype::Dropdown()` method allow to easily display a simple dropdown allowing the user to pick an item from a given itemtype.
This kind of dropdowns sadly doesn't support the "multiple" option of select2 which we can use on others dropdown like `showDropdownFromArray()`.
Creating a field that allow the user to select multiple items of the same itemtype is thus painful as you need to use a "dropdown from array", load the values manually, update data in multiple tables, ...

These changes add the 'multiple' option to the `$itemtype::Dropdown()` method to make this kind of fields easier to create.
There is also a new `getFieldsFromLinkTables` method in `CommonDBTM` that can be used to register such "multiple items" fields to allow GLPI to update them and load their values in the current item without having to write any code.

#### Working example from a plugin:
image

```php
$migration->createLinkTable("{$table}_users", EmailTarget::class, User::class);
$migration->addField("{$table}_users", "type", 'string');
$migration->createLinkTable("{$table}_groups", EmailTarget::class, Group::class);
$migration->addField("{$table}_groups", "type", 'string');
```

```php
public function getFieldsFromLinkTables(): array
{
   $base_table = self::getTable();
   $users_table = "{$base_table}_users";
   $groups_table = "{$base_table}_groups";

   return [
      "to_users"  => new FieldFromLinkTable(
         $users_table,
         User::class,
         ['type' => self::TARGET_TO]
      ),
      "to_groups" => new FieldFromLinkTable(
         $groups_table,
         Group::class,
         ['type' => self::TARGET_TO]
      ),
      "cc_users"  => new FieldFromLinkTable(
         $users_table,
         User::class,
         ['type' => self::TARGET_CC]
      ),
      "cc_groups" => new FieldFromLinkTable(
         $groups_table,
         Group::class,
         ['type' => self::TARGET_CC]
      ),
   ];
}
```
```twig
{{ fields.dropdownField(
   'User',
   'to_users',
   item.fields['to_users'],
   __('TO (users)'),
   {
      'right': 'all',
      'multiple' : true,
   }
) }}

{{ fields.dropdownField(
   'Group',
   'to_groups',
   item.fields['to_groups'],
   __('TO (groups)'),
   {
      'multiple' : true,
   }
) }}

{{ fields.dropdownField(
   'User',
   'cc_users',
   item.fields['cc_users'],
   __('CC (users)'),
   {
      'right': 'all',
      'multiple' : true,
   }
) }}

{{ fields.dropdownField(
   'Group',
   'cc_groups',
   item.fields['cc_groups'],
   __('CC (groups)'),
   {
      'multiple' : true,
   }
) }}
```

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets |

### CODE REVIEW

File: src/CommonDBTM.php [hunk 1]
```diff
@@ -434,6 +434,7 @@ public function getID()
    **/
     public function post_getFromDB()
```
Comment (cedric-anne) [inline src/CommonDBTM.php:435]: You should update all `post_getFromDB` overrides in order to call `parent::post_getFromDB()`, to ensure that this piece of code is always used.
Another solution would be to directly call this method in `CommonDBTM::getFromDB()`, right before the call to `$this->post_getFromDB()`.

File: src/CommonDBTM.php [hunk 2]
```diff
@@ -1442,6 +1443,8 @@ public function addNeededInfoToInput($input)
    **/
     public function prepareInputForAdd($input)
     {
+        $input = $this->handleMultipleItemsFieldsInInput($input);
```
Comment (cedric-anne) [inline src/CommonDBTM.php:1446]: Same remark.

File: src/CommonDBTM.php [hunk 3]
```diff
@@ -1812,6 +1815,8 @@ public function addMessageOnUpdateAction()
    **/
     public function prepareInputForUpdate($input)
     {
+        $input = $this->handleMultipleItemsFieldsInInput($input);
```
Comment (cedric-anne) [inline src/CommonDBTM.php:1818]: Same remark.

**Review by cedric-anne — COMMENTED** [review]
I started a review, then I figured out that storing multiple ids in database in a JSON-like format is a bad idea. Indeed, we need go towards the implementation of foreign key constraints in DB (to be able to prevent invalid/orphaned/... data in DB), and using this kind of format is not compatible with this.

Having ability to select multiple items from the same type is a good feature, but each selected item have to be stored in a relation table (like `glpi_computer_items`).

**Comment by AdrienClairembault** [comment]
> but each selected item have to be stored in a relation table 
I see, I'll look into it.

-> Converting back to draft meanwhile.

**Comment by AdrienClairembault** [comment]
Should be OK now, updated the original description.

**Comment by AdrienClairembault** [comment]
@orthagh 
I've removed the "autowiring" part.
Now the fields must be manually update from ` `post_addItem`, ` `post_updateItem` and `post_getFromDB`.

New plugin code:

```php
public function post_addItem()
{
   parent::post_addItem();

   $base_table = self::getTable();
   $linked_users_table = "{$base_table}_users";
   $linked_groups_table = "{$base_table}_groups";

   $this->update1NTableData(
      'to_users',
      User::class,
      $linked_users_table,
      ['type' => self::TARGET_TO]
   );
   $this->update1NTableData(
      'to_groups',
      Group::class,
      $linked_groups_table,
      ['type' => self::TARGET_TO]
   );
   $this->update1NTableData(
      'cc_users',
      User::class,
      $linked_users_table,
      ['type' => self::TARGET_CC]
   );
   $this->update1NTableData(
      'cc_groups',
      Group::class,
      $linked_groups_table,
      ['type' => self::TARGET_CC]
   );
}

public function post_updateItem($history = 1)
{
   parent::post_updateItem($history);

   $base_table = self::getTable();
   $linked_users_table = "{$base_table}_users";
   $linked_groups_table = "{$base_table}_groups";

   $this->update1NTableData(
      'to_users',
      User::class,
      $linked_users_table,
      ['type' => self::TARGET_TO]
   );
   $this->update1NTableData(
      'to_groups',
      Group::class,
      $linked_groups_table,
      ['type' => self::TARGET_TO]
   );
   $this->update1NTableData(
      'cc_users',
      User::class,
      $linked_users_table,
      ['type' => self::TARGET_CC]
   );
   $this->update1NTableData(
      'cc_groups',
      Group::class,
      $linked_groups_table,
      ['type' => self::TARGET_CC]
   );
}

public function post_getFromDB()
{
   parent::post_getFromDB();

   $base_table = self::getTable();
   $linked_users_table = "{$base_table}_users";
   $linked_groups_table = "{$base_table}_groups";

   $this->load1NTableData(
      'to_users',
      User::class,
      $linked_users_table,
      ['type' => self::TARGET_TO]
   );
   $this->load1NTableData(
      'to_groups',
      Group::class,
      $linked_groups_table,
      ['type' => self::TARGET_TO]
   );
   $this->load1NTableData(
      'cc_users',
      User::class,
      $linked_users_table,
      ['type' => self::TARGET_CC]
   );
   $this->load1NTableData(
      'cc_groups',
      Group::class,
      $linked_groups_table,
      ['type' => self::TARGET_CC]
   );
}
```

**Review by orthagh — APPROVED** [review]

**Comment by AdrienClairembault** [comment]
I've noticed a few PHP warning on update, wait for a fix.

**Comment by trasher** [comment]
New methods are not used at all in core, maybe tests should be added?

File: src/CommonDBTM.php [hunk 4]
```diff
@@ -1521,6 +1521,10 @@ public function update(array $input, $history = 1, $options = [])
                 $this->oldvalues = [];
 
                 foreach (array_keys($this->input) as $key) {
+                    if (is_array($this->input[$key])) {
+                        continue;
+                    }
```
Comment (cedric-anne) [inline src/CommonDBTM.php:1526]: I do not like this one. Unless I am wrong, it cannot trigger a warning unless the field exists in database, so why ignoring it ?

Comment (AdrienClairembault) [reply src/CommonDBTM.php:1526]: I have the following warning if I disable it:
```
[2022-01-17 11:04:38] glpiphplog.NOTICE:   *** PHP Notice (8): Array to string conversion in /home/adrie/localhost/gmaster/src/DBmysql.php at line 1263
  Backtrace :
  src/DBmysql.php:1393                               DBmysql::quoteValue()
  src/DBmysql.php:1427                               DBmysql->buildUpdate()
  src/CommonDBTM.php:608                             DBmysql->update()
  src/CommonDBTM.php:1597                            CommonDBTM->updateInDB()
  ...ncedforms/front/targets/emailtarget.form.php:55 CommonDBTM->update()
  ...ncedforms/front/targets/emailtarget.form.php:48 updateTarget()
  
```

**Comment by AdrienClairembault** [comment]
> New methods are not used at all in core, maybe tests should be added?

Great idea, done 👍

**Review by trasher — APPROVED** [review]

File: src/CommonDBTM.php (hunk 4 above)
Comment (cedric-anne) [reply src/CommonDBTM.php:1526]: The problem is that `if (array_key_exists($key, $this->fields)) {` is evaluated to true due to the fact that you recreate this "virtual" field in `post_getFromDB()` method of your item.

As update should only handle real db fields, I think the fix should be to do this
```diff
diff --git a/src/CommonDBTM.php b/src/CommonDBTM.php
index 90fc95df29..607d22079d 100644
--- a/src/CommonDBTM.php
+++ b/src/CommonDBTM.php
@@ -1521,7 +1521,7 @@ class CommonDBTM extends CommonGLPI
                 $this->oldvalues = [];
 
                 foreach (array_keys($this->input) as $key) {
-                    if (array_key_exists($key, $this->fields)) {
+                    if ($DB->fieldExists(static::getTable(), $key)) {
                       // Prevent history for date statement (for date for example)
                         if (
                             is_null($this->fields[$key])
```

Comment (AdrienClairembault) [reply src/CommonDBTM.php]: 👍  done.

File: src/CommonDBConnexity.php [hunk 5]
```diff
@@ -446,6 +446,9 @@ public function canConnexityItem(
    **/
     public function getHistoryChangeWhenUpdateField($field)
     {
+        if (is_array($this->fields[$field])) {
+            return null;
+        }
```
Comment (cedric-anne) [inline src/CommonDBConnexity.php:451]: This one should not be usefull anymore.

File: src/CommonDBTM.php [hunk 6]
```diff
@@ -5912,4 +5912,118 @@ public function getMassiveActionsForItem(): MassiveAction
...
+
+        // Compute item_2 params (linked objets(s))
+        $item_2_fk = getForeignKeyFieldForItemType($itemtype);
+
+        // Get input value
+        $input_value = $this->input[$field] ?? null;
+
+        // See dropdownField twig macro, needed for empty values as an empty
+        // array wont be sent in the HTML form
+        $input_defined_value = $this->input["_{$field}_defined"] ?? null;
```
Comment (cedric-anne) [inline src/CommonDBTM.php:5946]:
```suggestion
        $input_defined_value = (bool)$this->input["_{$field}_defined"] ?? false;
```

File: src/CommonDBTM.php [hunk 7]
```diff
@@ -5912,4 +5912,118 @@ public function getMassiveActionsForItem(): MassiveAction
...
+        if ($input_value === null && !$input_defined_value) {
+            return;
+        }
+
+        // Case 2: input was emptied -> remove all values
+        if (
+            ($input_value === null && $input_defined_value)
+            || (is_array($input_value) && ! count($input_value))
+        ) {
+            $DB->delete($table, $extra_input);
```
Comment (cedric-anne) [inline src/CommonDBTM.php:5958]: Even `CommonDBConnexity` items may have pre/post delete operations defined in their class, so this should be done using `CommonDBTM::delete()` method.

File: src/CommonDBTM.php [hunk 8]
```diff
@@ -5912,4 +5912,118 @@ public function getMassiveActionsForItem(): MassiveAction
...
+            return;
+        }
+
+        // Case 3: input was maybe modified -> delete missing values and add new ones
+        $DB->delete($table, array_merge($extra_input, [
```
Comment (cedric-anne) [inline src/CommonDBTM.php:5963]: Same remark here, should use `CommonDBTM::delete()`.

File: src/CommonDBTM.php [hunk 9]
```diff
@@ -5912,4 +5912,118 @@ public function getMassiveActionsForItem(): MassiveAction
...
+        foreach ($input_value as $item_2_id) {
+            if (in_array($item_2_id, $item_2_ids_db)) {
+                // Value exist
+                continue;
+            }
+
+            $DB->insert($table, array_merge($extra_input, [
+                $item_1_fk => $item_1_id,
+                $item_2_fk => $item_2_id
+            ]));
```
Comment (cedric-anne) [inline src/CommonDBTM.php:5988]: Even `CommonDBConnexity` items may have pre/post add operations defined in their class, so this should be done using `CommonDBTM::add()` method.

File: src/Dropdown.php [hunk 10]
```diff
@@ -131,31 +132,54 @@ public static function show($itemtype, $options = [])
...
         if (
-            $params['value'] !== null
+            isset($params['value'])
             && ((strlen($params['value']) == 0) || !is_numeric($params['value']) && $params['value'] != 'mygroups')
         ) {
             $params['value'] = 0;
         }
 
-        if (isset($params['toadd'][$params['value']])) {
+        if (!$params['multiple'] && isset($params['toadd'][$params['value']])) {
```
Comment (cedric-anne) [inline src/Dropdown.php:150]: Values from `toadd` should be added to `$names` on line 165.

File: src/Dropdown.php [hunk 11]
```diff
@@ -131,31 +132,54 @@ public static function show($itemtype, $options = [])
...
         }
 
         if ($params['readonly']) {
-            return self::getDropdownName($table, $params['value']);
+            if ($params['multiple']) {
+                // Multiple values, print the names separated by commas
+                $names = [];
+                foreach ($params['values'] as $value) {
+                    $names[] = self::getDropdownName($table, $value);
+                }
```
Comment (cedric-anne) [inline src/Dropdown.php:177]: `$names` are already computed in previous if/else block.

```suggestion
                // Multiple values, print the names separated by commas
```

File: src/Dropdown.php [hunk 12]
```diff
@@ -131,31 +132,54 @@ public static function show($itemtype, $options = [])
...
+                return implode(", ", $names);
+            } else {
+                // Single value, print the matching name
+                return self::getDropdownName($table, $params['value']);
```
Comment (cedric-anne) [inline src/Dropdown.php:181]: `$name` is already computed, I guess.
```suggestion
                return $name;
```

File: src/Dropdown.php [hunk 13]
```diff
@@ -181,10 +205,15 @@ public static function show($itemtype, $options = [])
         }
 
         if (!$item instanceof CommonTreeDropdown) {
-            $name = Sanitizer::unsanitize($name);
+            if ($params['multiple']) {
+                array_map(function ($name_to_unsanitize) {
+                    return Sanitizer::unsanitize($name_to_unsanitize);
+                }, $names);
```
Comment (cedric-anne) [inline src/Dropdown.php:211]: This method already handle arrays.
```suggestion
                $names = Sanitizer::unsanitize($names);
```

File: src/Html.php [hunk 14]
```diff
@@ -4683,6 +4684,7 @@ public static function jsAjaxDropdown($name, $field_id, $url, $params = [])
         $valuename = $params['valuename'];
         $on_change = $params["on_change"];
         $placeholder = $params['placeholder'] ?? '';
+        $multiple = $params['multiple'] ?? '';
```
Comment (cedric-anne) [inline src/Html.php:4687]: This param is defined in `$default_options`, so it is always defined. Thus `''` default value is not correct.
```suggestion
        $multiple = $params['multiple'];
```

File: src/Migration.php [hunk 15]
```diff
@@ -1561,4 +1561,35 @@ private function migrateSearchOptions()
...
+        string $class_2
+    ) {
+        $fk_1 = $class_1::getForeignKeyField();
+        $fk_2 = $class_2::getForeignKeyField();
+
+        $this->addPreQuery("
+            CREATE TABLE `$table` (
+                `id` int NOT NULL AUTO_INCREMENT,
+                `$fk_1` int NOT NULL DEFAULT '0',
+                `$fk_2` int NOT NULL DEFAULT '0',
```
Comment (cedric-anne) [inline src/Migration.php:1588]:
```suggestion
                `id` int unsigned NOT NULL AUTO_INCREMENT,
                `$fk_1` int unsigned NOT NULL DEFAULT '0',
                `$fk_2` int unsigned NOT NULL DEFAULT '0',
```

File: src/Migration.php [hunk 16]
```diff
@@ -1561,4 +1561,35 @@ private function migrateSearchOptions()
...
+                PRIMARY KEY (`id`),
+                KEY `$fk_1` (`$fk_1`),
+                KEY `$fk_2` (`$fk_2`)
+            ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci ROW_FORMAT=DYNAMIC;
```
Comment (cedric-anne) [inline src/Migration.php:1592]: Should use `DBConnection::getDefaultCharset()` and `DBConnection::getDefaultCollation()` values.

**Review by cedric-anne — CHANGES_REQUESTED** [review]

**Comment by AdrienClairembault** [comment]
@cedric-anne 
Last commit implement the requested changes.
Let me know if it's OK for you and then I'll rebase to fix the conflicts.

File: src/CommonDBTM.php [hunk 17]
```diff
@@ -5916,64 +5916,79 @@ public function getMassiveActionsForItem(): MassiveAction
...
+            $error = "$commondb_relation is not a CommonDBRelation item";
+            throw new InvalidArgumentException($error);
+        }
+
+        /** @var CommonDBRelation */
+        $commondb_relation = new $commondb_relation();
 
         // Compute item_1 params (current object)
-        $item_1_fk = getForeignKeyFieldForItemType(static::class);
+        $item_1_fk = $commondb_relation::$items_id_1;
```
Comment (cedric-anne) [inline src/CommonDBTM.php:5941]: Not really sure that main item is always the `$items_id_1` one.

Comment (cedric-anne) [reply src/CommonDBTM.php:5941]:
```php
if ($commondb_relation::$itemtype_1 === static::getType()) {
    // use 1
} elseif ($commondb_relation::$itemtype_2 === static::getType()) {
    // use 2
} elseif (preg_match('/^itemtype/', static::$itemtype_1) === 1 && preg_match('/^itemtype/', static::$itemtype_2) === 0) {
    // use 1
} elseif (preg_match('/^itemtype/', static::$itemtype_2) === 1 && preg_match('/^itemtype/', static::$itemtype_1) === 0) {
    // use 2
} else {
    // ambiguous, should fail
}
```

File: src/Migration.php [hunk 18]
```diff
@@ -1581,15 +1581,18 @@ public function createLinkTable(
...
 
+        $default_charset = DBConnection::getDefaultCharset();
+        $default_collation = DBConnection::getDefaultCollation();
+
         $this->addPreQuery("
             CREATE TABLE `$table` (
-                `id` int NOT NULL AUTO_INCREMENT,
-                `$fk_1` int NOT NULL DEFAULT '0',
-                `$fk_2` int NOT NULL DEFAULT '0',
+                `id` int unsigned NOT NULL AUTO_INCREMENT,
```
[the remark this answers is not in the document]
Comment (AdrienClairembault) [reply src/Migration.php:1589]: Ok, no changes to do here tho as it's not merged yet ?

[the remark this answers is not in the document]
Comment (cedric-anne) [reply src/Migration.php:1589]: It is now merged.

File: src/CommonDBTM.php [hunk 19]
```diff
@@ -5938,11 +5938,23 @@ protected function update1NTableData(
         $commondb_relation = new $commondb_relation();
 
         // Compute which item is item_1 and item_2
-        if (static::class == $commondb_relation::$itemtype_1) {
+        if (
+            static::class == $commondb_relation::$itemtype_1
+            || (
+                preg_match('/^itemtype/', $commondb_relation::$itemtype_1) === 1
+                && preg_match('/^itemtype/', $commondb_relation::$itemtype_2) === 0
+            )
```
Comment (cedric-anne) [inline src/CommonDBTM.php:5946]: This should be done after the `static::class == $commondb_relation::$itemtype_2` case.

File: src/Migration.php (hunk 18 above)
[the remark this answers is not in the document]
Comment (AdrienClairembault) [reply src/Migration.php:1590]: Done.

File: src/CommonDBTM.php (hunk 19 above)
Comment (AdrienClairembault) [reply src/CommonDBTM.php]: Done.

**Review by cedric-anne — APPROVED** [review]

**Comment by cedric-anne** [comment]
> Let me know if it's OK for you and then I'll rebase to fix the conflicts.

You can rebase now.

**Review by trasher — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Pull request merged**